### ISSUE TITLE
Testing: simulateHover and isOver interaction

### ISSUE THREAD

**Issue by gregj1** [gaearon/react-dnd#387]
TL;DR: Call simulateHover() twice in your tests if you're receiving unexpected results from monitor.isOver().

I spent some hours chasing this one down. I hope this note might save others from doing the same - its probably not a bug, just a gotcha for new players to be aware of.

I'm using a nested, sortable drag and drop component. I used the "Cancel on Outside Drop" and "greedy Drop Target" examples as guidance. I developed the component against a browser, got it working, and then set about writing test cases for it.

The core part of the test case was:
```
let source_component_id = source_component.getHandlerId();
let target_component_id = target_component.getHandlerId();
backend.simulateBeginDrag([source_component_id]);
backend.simulateHover([target_component_id]);
expect(onMoveSpy.calls.count()).toEqual(1);
```

Like the examples, my movespy should be called in the target's hover() function - after an isOver() check because of the nesting. The spy wasn't being called in the test, even though it appears to work fine in the browser.

After some debugging, it turns out that isOver is evaluating a list of target ids which is not being set until after the hover() function finishes. The relevant code, from `dnd-core/src/DragDropManager.js` is below. The target's hover() function is called by actionCreator.apply, the target ids are set by dispatch(action).

```
function bindActionCreator(actionCreator) {
      return function () {
        const action = actionCreator.apply(manager, arguments);
        if (typeof action !== 'undefined') {
          dispatch(action);
        }
      }
    }
```

Adding a second simulateHover to my test resolves the issue. The first pass will set the target ids in the dragOperation state, allowing the second hover() and isOver() calls to succeed. Probably the first call hover() is failing when running in the browser too, it's just being called again rapidly as the user moves the mouse.

```
let source_component_id = source_component.getHandlerId();
let target_component_id = target_component.getHandlerId();

backend.simulateBeginDrag([source_component_id]);
expect(monitor.getTargetIds()).toEqual([]);
backend.simulateHover([target_component_id]);
expect(onMoveSpy.calls.count()).toBe(0);
expect(monitor.getTargetIds()).toEqual([target_component_id]);
backend.simulateHover([target_component_id]); // targetIds are only available in the second call?
expect(onMoveSpy.calls.count()).toBe(1);
```

**Comment by gaearon** [comment]
Interesting, thanks for sharing. If you think this is a solveable problem (it might very well be) please send a pull request to https://github.com/gaearon/dnd-core. I don’t remember very well how things are wired here but I would suggest you to add a failing test similar to the one you wrote that expects `targetIds` to be available *inside* the `hover()` method, and then figure out how to make it pass (it might involve reordering how actions are dispatched inside `dnd-core` or something).
